vespalib/src/vespa/vespalib/util/file_area_freelist.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "file_area_freelist.h"
+#include <cassert>
+
+namespace vespalib::alloc {
+
+FileAreaFreeList::FileAreaFreeList()
+    : _free_areas(),
+      _free_sizes()
+{
+}
+
+FileAreaFreeList::~FileAreaFreeList() = default;
+
+void
+FileAreaFreeList::remove_from_size_set(uint64_t offset, size_t size)
+{
+    auto itr = _free_sizes.find(size);
+    assert(itr != _free_sizes.end());
+    auto &offsets = itr->second;
+    auto erased_count = offsets.erase(offset);
+    assert(erased_count != 0u);
+    if (offsets.empty()) {
+        _free_sizes.erase(itr);
+    }
+}
+
+uint64_t
+FileAreaFreeList::alloc(size_t size)
+{
+    auto sitr = _free_sizes.lower_bound(size);
```

Comment:
Consider making a helper function for lines 33-45, returning offset and old_size. I think this would make it easier to read this function.

Reply:
Done